Notebook

- Acquire data from the customers table in the mall_customers database.
- Summarize the data (include distributions and descriptive statistics).
- Detect outliers using IQR.
- Split data into train, validate, and test.
- Encode categorical columns using a one hot encoder (pd.get_dummies).
- Handles missing values.
- Scaling

Encapsulate your work in a wrangle_mall.py python module.

In [3]:
#basics
import numpy as np
import pandas as pd
import os


#viz
import matplotlib.pyplot as plt

#Data prep tools
from sklearn.model_selection import train_test_split
from itertools import filterfalse

#env
from env import host, username, password, get_db_url

##### Acquire Data

In [4]:
#sql query to use
sql = """
SELECT * FROM customers;
"""

In [7]:
#get data
df = pd.read_sql(sql,get_db_url('mall_customers'))
#general info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   customer_id     200 non-null    int64 
 1   gender          200 non-null    object
 2   age             200 non-null    int64 
 3   annual_income   200 non-null    int64 
 4   spending_score  200 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 7.9+ KB


In [10]:
#describe
df.describe().T

count    mean        std   min    25%    50%     75%    max
customer_id     200.0  100.50  57.879185   1.0  50.75  100.5  150.25  200.0
age             200.0   38.85  13.969007  18.0  28.75   36.0   49.00   70.0
annual_income   200.0   60.56  26.264721  15.0  41.50   61.5   78.00  137.0
spending_score  200.0   50.20  25.823522   1.0  34.75   50.0   73.00   99.0

spending score: 1 - 99
annual income: 15 to 137 >> seems like it may represent thousands

##### detect outliers using IQR

Function options:
- Take in whole dataframe
- All for a column to ignore parameter
  - EX: this has customer_id which is a PK.  May want to keep it, but needs to be ignored.
- Columns to find IQR for:
  - in case there are some 
- Want function to either:
  - Return informational dataframe with upper and lower bounds per column (infoonly)
  - Return a trimmed dataframe - and maybe another dataframe with what was done
  - Return a dataframe with new columns that identify if a row is an outlier
  
  
Better: Create a function that finds outliers and another that handles them

# PICKUP HERE
WORKING ON IQR OUTLIER FUNCTIONS.  NEED TO FIGURE OUT HOW TO FILTER OUT COLUMNS THAT ONLY CONTAIN ZEROS AND ONES.
CONSIDER FILTERFALSE (probably won't work)
CONSIDER LIST COMPREHENSION ( https://stackoverflow.com/questions/1207406/how-to-remove-items-from-a-list-while-iterating - not sure if index counts as list)

In [14]:
def only_0_or_1(ser):
    """
    Takes in a pandas series, returns True if it only contains zeros and ones.
    """
    return ser.isin([0,1]).all()

def get_iqr_outlier_bounds(df,include=None,exclude=None):
    """
    Returns dataframe with list of columns and the upper and lower bounds using the IQR method:
        LB = Q1 - 1.5 * IQR
        UB = Q3 + 1.5 * IQR
    If no columns passed to include or exclude, it defaults to finding outliers for all columns.
    Function will ignore non-numeric columns and columns that contain only 0s and 1s.
    
    Returns: Pandas Dataframe 
    Parameters:
           df: dataframe in which to find outliers
      include: list of columns to find outliers for
       excude: list of columns to NOT find outliers for.  Ignored if 'include'is set.   
    
    C88
    """
    #if include and exclude are None
    if not include and not exclude:
        columns = df.columns #returns index - iterable
    elif include in
    
    #check our columns are numeric
    columns = df[columns].select_dtypes(include='numeric')
    #check if only 0s and 1s
    for c in columns:
        #if series contains only zeros and ones
        if df[c].isin([0,1]).all():
            #

            
###IMPROVMENTS:
# 1) Currently only takes a list for include/exclude. Handle either single string input that is column name or list
# 2) if a numeric column only contains 0s and 1s, assume it to be boolean
# 3) If number of unique values is < than a certain number: 
#.     This will help identify if a column is really a category. Ruminate on this more
def handle_iqr_outliers(df,trim=False,include='all', exclude=None):
    """
    Takes in a dataframe and either trims outliers or creates column identifying outliers. 
    
    Outputs: None
    Returns: Pandas Dataframe
    Parameters:
      include, exclude: list
        
        
    
    
    """
    #must take in df, include/exclude, whether to trim outliers or create new columns
    
    #1 - get params
    
    #2 - get bound dataframe
    bounds = get_iqr_outlier_bounds(df,include,exclude)
    
    #3 -
    #If we want new column 
    if !trim:
        
    

NameError: name 'is_list_like' is not defined